src/compiler/srcloc.rs

```diff
@@ -29,19 +45,101 @@ pub struct Srcloc {
 
 impl Display for Srcloc {
     fn fmt(&self, formatter: &mut std::fmt::Formatter<'_>) -> Result<(), std::fmt::Error> {
-        match self.until {
+        match &self.until {
             None => formatter.write_str(&format!("{}({}):{}", self.file, self.line, self.col)),
-            Some((l, c)) => formatter.write_str(&format!(
+            Some(u) => formatter.write_str(&format!(
                 "{}({}):{}-{}({}):{}",
-                self.file, self.line, self.col, self.file, l, c
+                self.file, self.line, self.col, self.file, u.line, u.col
             )),
         }
     }
 }
 
 impl Srcloc {
+    pub fn new(name: Rc<String>, line: usize, col: usize) -> Self {
+        Srcloc {
+            file: name,
+            line,
+            col,
+            until: None,
+        }
+    }
+
+    pub fn ending(&self) -> Srcloc {
+        if let Some(u) = &self.until {
+            return Srcloc {
+                file: self.file.clone(),
+                line: u.line,
+                col: u.col,
+                until: None,
+            };
+        }
+        self.clone()
+    }
+
+    pub fn overlap(&self, other: &Srcloc) -> bool {
```

Comment:
Does `overlap(...) == true` mean "these two overlap", or does it mean "self is contained in other"?

Reply:
it means they overlap strictly.